controllers/cloud.redhat.com/clowdapp_reconciliation.go

```diff
@@ -0,0 +1,381 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	crd "github.com/RedHatInsights/clowder/apis/cloud.redhat.com/v1alpha1"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/clowderconfig"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/config"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/errors"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers"
+	provutils "github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers/utils"
+	rc "github.com/RedHatInsights/rhc-osdk-utils/resource_cache"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	core "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type ClowdAppReconciliation struct {
+	cache                 *rc.ObjectCache
+	recorder              record.EventRecorder
+	ctx                   *context.Context
+	client                client.Client
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
```

Comment:
Good catch re: public method. It is a hold over from an earlier design where this code was in its own package. Changed to private.

Its a method because you slapped my hand once for doing a literal declaration instead of using a constructor function :) There was probably a more subtle point to your asking me to use a function in the past, maybe like init logic or something, but it stuck in my head as "Pete wants me to use methods when creating instances of complex types. I can change it.

Reply:
Removed